### PR TITLE
dto used, sorting and filtering added

### PR DESCRIPTION
**Pull request by sharique-asl**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -30,6 +30,8 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-validation'
 	implementation 'org.springframework.boot:spring-boot-starter-data-redis' // giving us which type of database we are going to use for caching
 	implementation 'org.springframework.boot:spring-boot-starter-cache' // required for enabling caching mechanism
+	implementation 'org.modelmapper:modelmapper:3.2.0'
```
Comment (prototype201095) [inline build.gradle]: remove the dependency of the modelmapper

File: src/main/java/com/example/dbdemo/controller/CourseController.java [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 
-import com.example.dbdemo.model.*;
-import com.example.dbdemo.service.*;
+import com.example.dbdemo.model.Course;
+import com.example.dbdemo.service.CourseService;
+import com.example.dbdemo.dto.request.ResponseDTOWrapper;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.*;
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/CourseController.java]: use proper import statements

File: src/main/java/com/example/dbdemo/controller/CourseController.java [hunk 3]
```diff
@@ -17,41 +19,82 @@ public class CourseController {
...
+        return ResponseEntity
+                .status(HttpStatus.OK)
+                .body(
+                        ResponseDTOWrapper.<Course>builder()
+                                .items(courses)
+                                .build()
+                );
     }
 
     @GetMapping("/{id}")
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/CourseController.java]: merge the apis with getAllCourses()

File: src/main/java/com/example/dbdemo/controller/CourseController.java [hunk 4]
```diff
@@ -17,41 +19,82 @@ public class CourseController {
...
+        return ResponseEntity
+                .status(HttpStatus.OK)
+                .body(
+                        ResponseDTOWrapper.<Course>builder()
+                                .items(courses)
+                                .build()
+                );
     }
 
     @PostMapping
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/CourseController.java]: add path for the creating course, please add content type also.

File: src/main/java/com/example/dbdemo/controller/CourseController.java [hunk 5]
```diff
@@ -17,41 +19,82 @@ public class CourseController {
...
-    public ResponseEntity<Course> createCourse(@RequestBody Course course) {
+    public ResponseEntity<ResponseDTOWrapper<Course>> createCourse(@RequestBody Course course) {
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/CourseController.java]: create a string as a response message instead of returning the entire object

File: src/main/java/com/example/dbdemo/controller/CourseController.java [hunk 6]
```diff
@@ -17,41 +19,82 @@ public class CourseController {
...
+                .status(updatedCourse != null ? HttpStatus.OK : HttpStatus.NOT_FOUND)
+                .body(
+                        ResponseDTOWrapper.<Course>builder()
+                                .items(updatedCourse != null ? Collections.singletonList(updatedCourse) : null)
+                                .errorMessage(updatedCourse == null ? "Course to be updated not found" : null)
+                                .build()
+                );
     }
 
+    @DeleteMapping("/{id}")
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/CourseController.java]: give the convenient name for the api(s)

File: src/main/java/com/example/dbdemo/controller/DepartmentController.java [hunk 7]
```diff
@@ -17,38 +19,83 @@ public class DepartmentController {
     private DepartmentService departmentService;
 
     @GetMapping
-    public ResponseEntity<List<Department>> getAllDepartments() {
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/DepartmentController.java:20]: merge all fetch department details API(s) into one

File: src/main/java/com/example/dbdemo/controller/StudentController.java [hunk 8]
```diff
@@ -1,100 +1,224 @@
...
-        return ResponseEntity.ok(students);
+    public ResponseEntity<ResponseDTOWrapper<StudentResponseDTO>> getAllStudents(
+            @RequestParam(name = "name", required = false) String name,
+            @RequestParam(name = "gender", required = false) Gender gender,
+            @RequestParam(name = "sort", required = false) String sort
+    ) {
+        try {
+            List<Student> students = null;
+
+            if (sort != null && !sort.isEmpty()) {
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/StudentController.java]: make a common utility function for the validations

File: src/main/java/com/example/dbdemo/service/DepartmentServiceImpl.java [hunk 9]
```diff
@@ -24,17 +27,21 @@ public Department getDepartmentById(Long id) {
     }
 
     @Override
-    public List<Department> getDepartmentsByIds(List<Long> ids) {
+    public List<Department> getDepartmentsByIds(@NotNull List<Long> ids) {
+
+        if (ids.isEmpty())
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/service/DepartmentServiceImpl.java:29]: create a utility function and use it for validations

File: src/main/java/com/example/dbdemo/service/StudentServiceImpl.java [hunk 10]
```diff
@@ -1,46 +1,73 @@
...
     @Override
-    @Cacheable(value = "student", key = "#id")
-//    @Transactional
+//    @Cacheable(value = "student", key = "#id")
     public Student getStudentById(Long id) {
         return studentRepository.findById(id).orElse(null);
     }
 
+    public List<Student> getFilteredStudents(String name, Gender gender) {
+        List<Student> students = studentRepository.findAll();
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/service/StudentServiceImpl.java]: instead of finding all the student details from the DB, can we write a separate  DB query that will return filtered students?

File: src/main/java/com/example/dbdemo/controller/DepartmentController.java (hunk 7 above)
Comment (shekher9) [reply src/main/java/com/example/dbdemo/controller/DepartmentController.java:20]: Add the API name for all the places where it's empty
it's recommended to give some api paths for all the api's
Thanks

File: src/main/java/com/example/dbdemo/service/StudentServiceImpl.java (hunk 10 above)
Comment (sharique-asl) [reply src/main/java/com/example/dbdemo/service/StudentServiceImpl.java]: Created a custom query method in StudentRepository

File: src/main/java/com/example/dbdemo/controller/CourseController.java [hunk 11]
```diff
@@ -1,57 +1,98 @@
...
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
 
+
+import java.util.Collections;
 import java.util.List;
 
 @RestController
-@RequestMapping("/courses")
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/CourseController.java:13]: why did you remove class level request mapping ?

File: src/main/java/com/example/dbdemo/controller/FacultyController.java [hunk 12]
```diff
@@ -1,46 +1,165 @@
...
+                                    .build()
+                    );
+        }
     }
 
-    @PostMapping("/faculties")
-    public ResponseEntity<Faculty> createFaculty(@RequestBody Faculty faculty) {
-        Faculty createdFaculty = facultyService.createFaculty(faculty);
-        return new ResponseEntity<>(createdFaculty, HttpStatus.CREATED);
+    @GetMapping("/faculties/get/ids")
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/FacultyController.java]: please add produces and consumes type for all the API(s)

File: src/main/java/com/example/dbdemo/controller/CourseController.java (hunk 11 above)
Comment (protik4alumnux) [reply src/main/java/com/example/dbdemo/controller/CourseController.java:13]: This is for consistency. We removed it to have to explicitly mentioned over the get, post, delete and patch methods such as:
> ```@GetMapping("/courses/get")```

To make it more readable

File: build.gradle [hunk 13]
```diff
@@ -30,6 +30,7 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-validation'
 	implementation 'org.springframework.boot:spring-boot-starter-data-redis' // giving us which type of database we are going to use for caching
 	implementation 'org.springframework.boot:spring-boot-starter-cache' // required for enabling caching mechanism
+
```
Comment (shekher9) [inline build.gradle]: remove unwanted spaces

File: src/main/java/com/example/dbdemo/controller/CourseController.java [hunk 14]
```diff
@@ -1,57 +1,98 @@
...
 public class CourseController {
     // Course CRUD operations
     @Autowired
     private CourseService courseService;
 
-    @GetMapping
-    public ResponseEntity<List<Course>> getAllCourses() {
+    @GetMapping("/courses/get")
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/controller/CourseController.java]: write something like "/courses/getAllCourses"

File: src/main/java/com/example/dbdemo/controller/CourseController.java [hunk 15]
```diff
@@ -1,57 +1,98 @@
...
+                                .items(courses)
+                                .build()
+                );
     }
 
-    @PostMapping
-    public ResponseEntity<Course> createCourse(@RequestBody Course course) {
-        Course createdCourse = courseService.createCourse(course);
-        return new ResponseEntity<>(createdCourse, HttpStatus.CREATED);
+    @PostMapping("/courses/add")
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/controller/CourseController.java]: "/addCourse"

File: src/main/java/com/example/dbdemo/controller/CourseController.java [hunk 16]
```diff
@@ -1,57 +1,98 @@
...
+        } catch (Exception e) {
+            return ResponseEntity
+                    .status(HttpStatus.INTERNAL_SERVER_ERROR)
+                    .body("Failure");
+        }
     }
 
-    @PutMapping("/{id}")
-    public ResponseEntity<Course> updateCourse(@PathVariable Long id, @RequestBody Course course) {
+    @PutMapping("courses/edit/{id}")
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/controller/CourseController.java]: "/updateById/{id}"

File: src/main/java/com/example/dbdemo/controller/StudentController.java [hunk 17]
```diff
@@ -1,100 +1,190 @@
...
-
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RequestBody;
+
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/controller/StudentController.java]: remove unwanted spaces

File: src/main/java/com/example/dbdemo/controller/StudentController.java [hunk 18]
```diff
@@ -1,100 +1,190 @@
...
+    @DeleteMapping("/students/delete/{id}")
+    public ResponseEntity<ResponseDTOWrapper<String>> deleteStudent(@PathVariable Long id) {
+        try {
+            Boolean status = studentService.deleteStudent(id);
+
+            return ResponseEntity
+                    .status(status ? HttpStatus.OK : HttpStatus.BAD_REQUEST)
+                    .body(
+                            ResponseDTOWrapper.<String>builder()
+                                    .items(Collections.singletonList(status ? "Deletion successful" : "Deletion unsuccessful"))
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/controller/StudentController.java]: Use deletion failed instead of deletion unsuccessful

File: src/main/java/com/example/dbdemo/dto/request/CreateFacultyRequestDTO.java [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDate;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class CreateFacultyRequestDTO {
+    @NotNull
+    private Long id;
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/dto/request/CreateFacultyRequestDTO.java]: Please remove the id from all the createRequestDTO's

File: src/main/java/com/example/dbdemo/dto/request/CreateStudentRequestDTO.java [hunk 20]
```diff
@@ -1,12 +1,117 @@
 package com.example.dbdemo.dto.request;
 
+import com.example.dbdemo.dto.response.StudentResponseDTO;
+import com.example.dbdemo.model.Student;
+import com.example.dbdemo.utilities.Gender;
+import jakarta.persistence.*;
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/dto/request/CreateStudentRequestDTO.java]: use full import statements

File: src/main/java/com/example/dbdemo/dto/request/UpdateFacultyRequestDTO.java [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDate;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class UpdateFacultyRequestDTO {
+    private Long id;
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/dto/request/UpdateFacultyRequestDTO.java:25]: While doing update id should be mandatory

File: src/main/java/com/example/dbdemo/dto/response/DepartmentResponseDTO.java [hunk 22]
```diff
@@ -1,17 +1,44 @@
...
+import lombok.NoArgsConstructor;
 
 @Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
 public class DepartmentResponseDTO {
     private Long departmentId;
+
+    @NotBlank(message = "Department Name should not be null or empty")
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/dto/response/DepartmentResponseDTO.java]: pls remove all the validations from all the response classes

File: src/main/java/com/example/dbdemo/model/Faculty.java [hunk 23]
```diff
@@ -3,85 +3,61 @@
...
-    @NotBlank
-    @Pattern(regexp = "^[a-zA-Z0-9]*$", message = "PAN should be alphanumeric with no special characters.")
-    @Size(min = 10, max = 10, message = "PAN should be 10 characters.")
     private String pan;
 
     private String maritalStatus;
 
     @PastOrPresent(message = "Date of joining should be in the past or present.")
+    @Builder.Default
     private LocalDate dateOfJoining = LocalDate.now();
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/model/Faculty.java]: pls check the format only need the date part, not all of yyyy-mm-ddThh:mm:sec.microsec
you can use Date also

File: src/main/java/com/example/dbdemo/model/Student.java [hunk 24]
```diff
@@ -13,73 +16,48 @@
...
-    @Size(min = 12, max = 12, message = "Aadhar number should be 12 digits.")
-    @Pattern(regexp = "^[0-9]*$", message = "Aadhar number should not contain special characters.")
     private String aadharNumber;
 
-    @NotBlank
-    @Pattern(regexp = "^[a-zA-Z0-9]*$", message = "PAN should be alphanumeric with no special characters.")
-    @Size(min = 10, max = 10, message = "PAN should be 10 characters.")
     private String pan;
+
+    @Builder.Default
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/model/Student.java]: No need to use the @Builder.Default

File: src/main/java/com/example/dbdemo/service/StudentServiceImpl.java [hunk 25]
```diff
@@ -1,46 +1,76 @@
...
+//    @Cacheable(value = "student", key = "#id")
     public Student getStudentById(Long id) {
         return studentRepository.findById(id).orElse(null);
     }
 
+    public List<Student> getFilteredStudents(String name, Gender gender) {
+
+        return studentRepository.findFilteredStudents(name, gender);
+
+//        List<Student> students = studentRepository.findAll();
```
Comment (shekher9) [inline src/main/java/com/example/dbdemo/service/StudentServiceImpl.java]: remove all the commented codes

**Review by shekher9 — COMMENTED** [review]
Need Work

File: src/main/java/com/example/dbdemo/controller/DepartmentController.java [hunk 26]
```diff
@@ -2,53 +2,87 @@
...
+                                .errorMessage(updatedDepartment == null ? "Department to be updated not found" : null)
+                                .build()
+                );
     }
 
-    @DeleteMapping("/{id}")
-    public ResponseEntity<Void> deleteDepartment(@PathVariable Long id) {
-        departmentService.deleteDepartment(id);
-        return ResponseEntity.noContent().build();
+    @DeleteMapping(value = "/deleteDepartment/{id}", produces = MediaType.APPLICATION_JSON_VALUE, consumes = MediaType.ALL_VALUE)
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/DepartmentController.java]: remove consumes for delete operations

File: src/main/java/com/example/dbdemo/controller/FacultyController.java [hunk 27]
```diff
@@ -1,46 +1,159 @@
...
-        return ResponseEntity.ok(faculties);
-    }
 
-    @GetMapping("/faculties/{id}")
-    public ResponseEntity<Faculty> getFacultyById(@PathVariable Long id) {
-        Faculty faculty = facultyService.getFacultyById(id);
-        return faculty != null ? ResponseEntity.ok(faculty) : ResponseEntity.notFound().build();
+    @PostMapping(value = "/getAllFaculties", produces = MediaType.APPLICATION_JSON_VALUE, consumes = MediaType.ALL_VALUE)
+    public ResponseEntity<ResponseDTOWrapper<FacultyResponseDTO>> getAllFaculties() {
+        try {
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/FacultyController.java:44]: add the filters(i.e. name and gender) for the students

File: src/main/java/com/example/dbdemo/controller/FacultyController.java [hunk 28]
```diff
@@ -1,46 +1,159 @@
...
+                    );
+        }
     }
 
-    @PostMapping("/faculties")
-    public ResponseEntity<Faculty> createFaculty(@RequestBody Faculty faculty) {
-        Faculty createdFaculty = facultyService.createFaculty(faculty);
-        return new ResponseEntity<>(createdFaculty, HttpStatus.CREATED);
+    @GetMapping(value = "/getFacultiesByIds", produces = MediaType.APPLICATION_JSON_VALUE, consumes = MediaType.ALL_VALUE)
+    public ResponseEntity<ResponseDTOWrapper<FacultyResponseDTO>> getFacultiesByIds(@RequestParam List<Long> id) {
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/controller/FacultyController.java]: remove consumes from requestmapping

File: src/main/java/com/example/dbdemo/model/Faculty.java [hunk 29]
```diff
@@ -1,87 +1,70 @@
...
-    @Size(min = 10, max = 10, message = "PAN should be 10 characters.")
     private String pan;
 
     private String maritalStatus;
 
     @PastOrPresent(message = "Date of joining should be in the past or present.")
-    private LocalDate dateOfJoining = LocalDate.now();
+    private Date dateOfJoining = new Date();
 
+    @Builder.Default
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/model/Faculty.java]: remove Builder.default

File: src/main/java/com/example/dbdemo/service/CourseServiceImpl.java [hunk 30]
```diff
@@ -24,24 +27,35 @@ public Course getCourseById(Long id) {
...
+    }
+
+    @Override
+    public Course createCourse(@Valid Course course) {
         return courseRepository.save(course);
     }
 
     @Override
-    public Course updateCourse(Long id, Course course) {
+    public Course updateCourse(@NotNull @Min(1) Long id, @Valid Course course) {
```
Comment (prototype201095) [inline src/main/java/com/example/dbdemo/service/CourseServiceImpl.java]: 1. put `@valid` annotation in the controller (also check for the other API(s) as well)
2. No need to take  id explicitly as it is already there in the request payload

**Review by prototype201095 — COMMENTED** [review]
Please check the suggestion to the entire codebase

**Review by shekher9 — APPROVED** [review]
